WIP: Adds race-condition-proof plugin booting and handling of new DOM content

mrclay opened the pull request:
This introduces "plugin boot" modules which, if present, are automatically required by a new `elgg/booted` module. This is a solution to #7926 (race conditions with plugin hook registrations).

As well, this adds a "behaviors" system modeled after [Drupal's](https://www.lullabot.com/blog/article/understanding-javascript-behaviors-drupal). If you have to transform DOM content (can't rely on event delegation), you write your boot module to return a "attachBehavior" function, and if all scripts properly use `elgg.ajax.attachBehaviors()`, then your transformations will be applied to XHR-loaded content. Yes, this is a convention that devs will have to follow, but it works.

I applied this to enough stuff in plugins and core stuff to get an idea for how this works.
- [x] test in other browsers (particularly async require inside definition)
- [x] fix AudioPlayer.embed() in IE9
- [x] complain if boot module returns non-object or unexpected keys.
- [x] change standard boot module to `boot/$plugin_id` to avoid NS collisions
- [ ] docs
- [ ] after 1.x merged, attach behaviors in elgg.Comment save action
- [ ] after #8993 is merged, rebase this to just add behaviors

mrclay committed:
feature(javascript): Add plugin boot modules and element “behaviors”

Plugins can provide a <plugin_id>/boot AMD module which will be automatically
loaded by a new elgg/booted module. Plugin boot modules are the ideal
place to register plugin hooks, and the return values can be used to add
behavior to DOM elements (e.g. turn textareas into WYSIWYGs).

The elgg/booted module can be required to make sure all the plugin hooks
and behaviors are registered.

The final system hooks are now triggered in elgg/booted.

Fixes #7926, #6261

mrclay committed:
chore(plugins): Convert zaudio to boot module/behaviors system

mrclay committed:
chore(plugins): Convert ckeditor to boot module/behaviors system

mrclay committed:
chore(plugins): Convert likes to boot module system

mrclay committed:
chore(comments): Convert comments JS to AMD

mrclay committed:
chore(views): Remove unneeded external view registration (file is .js)

mrclay commented:
cc @ewinslow @hypeJunction

mrclay commented:
This really is 2 independent features. The boot module I think is a solid solution to hooks race conditions, but the "behaviors" hoops we have to jump through.. Inline `require()` calls in views sure look simpler.

mrclay commented on views/default/js/elgg/booted.js.php:
```diff
@@ -0,0 +1,40 @@
+<?php
+
+$modules = [];
+foreach (elgg_get_plugins() as $plugin) {
```
Probably should be priority order.

hypeJunction commented:
Seems solid. I like the approach.

hypeJunction commented:
Seems solid. I like the approach.

mrclay commented on mod/zaudio/views/default/js/zaudio/boot.js:
```diff
@@ -0,0 +1,16 @@
+define(function (require) {
+	var $ = require('jquery');
+
+	return {
```
It might be easier for users to create these modules correctly if they were required to return objects and the booted module complained if unexpected keys were present. Otherwise a typo would cause code to silently not run.

hypeJunction replied:
Perhaps we could use a class for this?

mrclay replied:
A type would mean users have to add another dependency to all boot modules and you still have to check unexpected keys, but it might get still make more sense to users.

hypeJunction replied:
It could be bundled with JS lib. 

```js
var BootManager = function() {};
BootManager.prototype = {
   addBehavior: function(closure),
};
```

mrclay commented on mod/zaudio/views/default/js/zaudio/boot.js:
```diff
@@ -0,0 +1,16 @@
+		addBehavior: function (context) {
+			var $targets = $('[data-zaudio-player]', context);
+			if ($targets.length) {
+				require(['elgg/zaudio'], function (zaudio) {
```
I was a bit surprised this works (async require inside definition) but it works fine, allowing the optional loading of that set of dependencies only if we need it. In this PR I do the same for CKeditor.

mrclay commented:
I'm open to renaming any of the APIs/modules added here.

hypeJunction commented on engine/lib/elgglib.php:
```diff
@@ -1906,7 +1906,9 @@ function _elgg_init() {
 	
-	elgg_register_external_view('js/elgg/UserPicker.js', true);
+	elgg_register_simplecache_view('js/elgg/booted.js');
+	elgg_require_js('elgg/Comment');
```
Can this be required only if needed?

mrclay committed:
chore(comments): only require comments AMD module where needed

ewinslow commented on js/lib/ajax.js:
```diff
@@ -106,6 +106,21 @@ elgg.ajax.handleAjaxError = function(xhr, status, error) {
+ *
+ * @param {jQuery} $context jQuery-wrapped DOM element(s) in which to attach behaviors
+ */
+elgg.ajax.attachBehaviors = function($context) {
```
Move this to its own `elgg/ajax/attachBehaviors` AMD module?

ewinslow replied:
Or elgg/behaviors/attach or whatever

mrclay replied:
This was going to be a convenience function, but we don't need it.

ewinslow commented on mod/ckeditor/views/default/js/ckeditor/boot.js:
```diff
@@ -0,0 +1,16 @@
+define(function (require) {
```
Should this be `elgg/ckeditor/boot` since it is not written by the ckeditor team? Feels strange to squat on their namespace with our own code.

ewinslow replied:
Doing so would match the elgg/ckeditor/config module below, too.

mrclay replied:
The convention this introduces automatically requires `$plugin_id/boot`. We could change the convention to something like `boot/$plugin_id`?

ewinslow commented on mod/ckeditor/views/default/js/ckeditor/boot.js:
```diff
@@ -0,0 +1,16 @@
+		addBehavior: function (context) {
+			$longtexts = $('.elgg-input-longtext:not([data-cke-init])', context);
+			if ($longtexts.length) {
+				require(['elgg/ckeditor'], function(elggCKEditor) {
```
Kinda wish there were a promise-based API for this so you can put all requires at top but not block on them:

```js
var $ = require('jquery'); // block module executing on jquery
var elggCKEditorPromise = require(['elgg/ckeditor']); // don't block, but start loading now.

// ...
elggCKEditorPromise.then(function(elggCKEditor) { // at the point we need it, wait for promise to resolve first
  $longtexts.ckeditor(elggCKEditor.init, elggCKEditor.config);
});
```

ewinslow replied:
@jrburke

mrclay replied:
If the promise gets resolved before `then` is called, would it still work or would that be a race?

mrclay replied:
Anyway the point of this is to conditionally avoid loading the module completely.

mrclay committed:
chore(javascript): More boot module design changes

Boot modules are named “boot/$plugin_id”
Boot modules must return an instance of elgg/Plugin
Removed elgg.ajax.attachBehavior helper function
Always require elgg/Comments (in case added to page)
Makes sure behaviors added in plugin order

mrclay commented:
In the latest commit:

Boot modules are named `boot/$plugin_id`
Boot modules must return an instance of elgg/Plugin
Removed elgg.ajax.attachBehavior helper function
Always require `elgg/Comments` (in case comments are added to the page)

mrclay commented:
On IE9 AudioPlayer.embed() silently fails, no errors, need to get unminified source. Works on 1.11.

mrclay commented:
I've submitted a version of this without the behaviors module #8365 since it's best to consider these independently.

mrclay commented:
Replaced by #9810
